### ISSUE TITLE
Class Loading Deadlock / Race Condition Involving DSL.using

### ISSUE THREAD

**Issue by randombk** [jOOQ/jOOQ#13814]
### Expected behavior
Application should not hang when running `DSL.using()` from two codepaths. The actual production breakage involved running two `DSL.using()` triggered from two codepaths: one manual invocation alongside one triggered by classloading a generated jOOQ Table class (See "Actual Production Stacktrace" example below). However, the issue can also be triggered using the example code below. 

### Actual behavior
Deadlock occurs with the following example stack trace:

```
"Thread-2@1014" prio=5 tid=0x13 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at org.jooq.impl.AbstractDataType.getTypeName(AbstractDataType.java:527)
	  at org.jooq.impl.DefaultDataType.getTypeName(DefaultDataType.java:136)
	  at org.jooq.impl.ArrayDataType.<init>(ArrayDataType.java:63)
	  at org.jooq.impl.AbstractDataType.getArrayDataType(AbstractDataType.java:611)
	  at org.jooq.impl.DefaultDataType.getArrayDataType(DefaultDataType.java:136)
	  at org.jooq.impl.DefaultDataType.<init>(DefaultDataType.java:392)
	  at org.jooq.impl.DefaultDataType.<init>(DefaultDataType.java:335)
	  at org.jooq.impl.DefaultDataType.<init>(DefaultDataType.java:331)
	  at org.jooq.impl.DefaultDataType.<init>(DefaultDataType.java:315)
	  at org.jooq.impl.BuiltInDataType.<init>(BuiltInDataType.java:60)
	  at org.jooq.impl.SQLDataType.<clinit>(SQLDataType.java:161)
	  at java.lang.Class.forName0(Class.java:-1)
	  at java.lang.Class.forName(Class.java:375)
	  at org.jooq.impl.DefaultDSLContext.<clinit>(DefaultDSLContext.java:5234)
	  at org.jooq.impl.DSL.using(DSL.java:507)
	  at archivia.Test.lambda$main$1(Test.java:23)
	  at archivia.Test$$Lambda$28/0x0000000800086428.run(Unknown Source:-1)
	  at java.lang.Thread.run(Thread.java:833)

"Thread-1@1013" prio=5 tid=0x12 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at org.jooq.impl.DSL.using(DSL.java:886)
	  at org.jooq.impl.Tools.<clinit>(Tools.java:3592)
	  at java.lang.Class.forName0(Class.java:-1)
	  at java.lang.Class.forName(Class.java:375)
	  at archivia.Test.lambda$main$0(Test.java:16)
	  at archivia.Test$$Lambda$27/0x0000000800086208.run(Unknown Source:-1)
	  at java.lang.Thread.run(Thread.java:833)
```

### Steps to reproduce the problem
```java
import org.jooq.SQLDialect;
import org.jooq.impl.DSL;

public class Test {
    public static void main(String... args) {
        Thread[] threads = {
                new Thread(() -> {
                    try {
                        Class.forName("org.jooq.impl.Tools");
                    } catch (Throwable e1) {
                        e1.printStackTrace();
                    }
                }),
                new Thread(() -> {
                    try {
                        DSL.using(SQLDialect.POSTGRES);
                    } catch (Throwable e) {
                        e.printStackTrace();
                    }
                })
        };
        threads[0].start();
        threads[1].start();
    }
}

```

### Actual production stacktrace (showing non-synthetic codepath)

```
"eventLoopGroupProxy-5-1@6664" daemon prio=5 tid=0x20 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at org.jooq.impl.AbstractDataType.getTypeName(AbstractDataType.java:527)
	  at org.jooq.impl.DefaultDataType.getTypeName(DefaultDataType.java:135)
	  at org.jooq.impl.ArrayDataType.<init>(ArrayDataType.java:63)
	  at org.jooq.impl.AbstractDataType.getArrayDataType(AbstractDataType.java:611)
	  at org.jooq.impl.DefaultDataType.getArrayDataType(DefaultDataType.java:135)
	  at org.jooq.impl.DefaultDataType.<init>(DefaultDataType.java:391)
	  at org.jooq.impl.DefaultDataType.<init>(DefaultDataType.java:334)
	  at org.jooq.impl.DefaultDataType.<init>(DefaultDataType.java:330)
	  at org.jooq.impl.DefaultDataType.<init>(DefaultDataType.java:314)
	  at org.jooq.impl.BuiltInDataType.<init>(BuiltInDataType.java:60)
	  at org.jooq.impl.SQLDataType.<clinit>(SQLDataType.java:161)
	  at java.lang.Class.forName0(Class.java:-1)
	  at java.lang.Class.forName(Class.java:375)
	  at org.jooq.impl.DefaultDSLContext.<clinit>(DefaultDSLContext.java:5129)
	  at org.jooq.impl.DSL.using(DSL.java:772)
	  at archivia.ArchiviaServerContext.getSQLContext(ArchiviaServerContext.kt:148)
	  [...]

"eventLoopGroupProxy-5-2@7020" daemon prio=5 tid=0x22 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at org.jooq.impl.DSL.using(DSL.java:886)
	  at org.jooq.impl.Tools.<clinit>(Tools.java:3579)
	  at org.jooq.impl.FieldsImpl.<init>(FieldsImpl.java:85)
	  at org.jooq.impl.TableImpl.<init>(TableImpl.java:245)
	  at org.jooq.impl.TableImpl.<init>(TableImpl.java:230)
	  at archivia.db.wmds.tables.Wmds_00003Collection.<init>(Wmds_00003Collection.java:113)
	  at archivia.db.wmds.tables.Wmds_00003Collection.<init>(Wmds_00003Collection.java:109)
	  at archivia.db.wmds.tables.Wmds_00003Collection.<init>(Wmds_00003Collection.java:134)
	  at archivia.db.wmds.tables.Wmds_00003Collection.<clinit>(Wmds_00003Collection.java:42)
	  at archivia.db.wmds.Tables.<clinit>(Tables.java:28)
	  at archivia.wmds.nodes.ArchiviaNodeTypes.<clinit>(ArchiviaNodeTypes.kt:16)
	  [...]
```

### Versions

- jOOQ: `3.17.2`. Issue did not occur on `3.15.5`.
- Java: OpenJDK 18.0.1.1
- Database (include vendor): N/A
- OS: Linux `Linux 5.18.11 x86_64`
- JDBC Driver (include name if unofficial driver): N/A

**Comment by randombk** [comment]
Temporary workaround is to put a `DSL.using(SQLDialect.[...])` in the main function before any threads are created.

**Comment by lukaseder** [comment]
Thanks for your report. I'll look into this, soon

**Comment by lukaseder** [comment]
A related issue has surfaced: https://github.com/jOOQ/jOOQ/issues/14769

**Comment by lukaseder** [comment]
https://github.com/jOOQ/jOOQ/issues/14769 has been fixed in 3.18.1 and 3.17.10. Can you confirm this also solves your issue?

**Comment by lukaseder** [comment]
Closing this as a duplicate of https://github.com/jOOQ/jOOQ/issues/14769.

It may or may not be the same problem, hard to say. I'll be happy to review a reproducer, should the issue persist.
